mozilla/fx-private-relay
# MPP-3827: Move Profile and RegisteredSubdomain models to privaterelay

This PR moves Django models and related code from `emails` to `privaterelay`. 

It does _not_ update the database. The moved models have the [`Meta.db_table` property](https://docs.djangoproject.com/en/4.2/topics/db/models/#meta-options) set to ensure the table names do not change (still `emails_profile` and `emails_registeredsubdomain`. The migrations use the [SeparateDatabaseAndState](https://docs.djangoproject.com/en/4.2/ref/migration-operations/#separatedatabaseandstate) operation to avoid SQL.  Deployment will work, and that the terraform SQL will continue to work. Thanks to https://davit.hashnode.dev/django-move-model for explaining this technique.

It may be possible to smoothly rename the tables as well. PostgreSQL [views](https://www.postgresql.org/docs/current/sql-createview.html) are updatable, so we could create a `emails_profile` view that allows existing code to continue to query, insert, and update to `emails_profile` while the table is named `privaterelay_profile`. There are two issues. One is the SQLite does not have updatable views, so I'd need to craft triggers to make tests pass. The second is that if a test needs to run in a transaction, `TRUNCATE` is used and further complicates things. Both could be handled. SQLite triggers are possible to write, just hard. I eliminated all the transaction tests, and an upstream contribution to `pytest_django` is possible. We could also just say "dev and migration tests are broken for a bit". In any case, I'd like to avoid any table renames in this PR.

* Models: 
  - Moved `Profile`, since it contains a mix of email, phone, and account information for a Relay user
  - Moved `RegisteredSubdomain`, since it is concerned with `Profile.subdomain`, with a validator that uses it. I could also argue that this should remain in `emails`, since subdomains are an email feature.
  - Did not move `AbuseMetrics`, since it only has email usage and doesn't have a relationship to the `Profile`. I could also argue that this should move to `privaterelay`, since it could gain phone abuse metrics in the future.
* Validators:
  - Moved `valid_available_subdomain` to `privaterelay`, to follow `RegisteredSubdomain`
* Signals:
  - Moved all signals to `privaterelay`, to follow `Profile`
* Exceptions:
  - Moved `CannotMakeSubdomainException` to `privaterelay`, to follow `RegisteredSubdomain`
* Tests, Admin - also moved to stay with related code  

# How to test

* Get a hot beverage. 
* Review the code.
* Think about these questions:
   - Should we instead break up `Profile` into `RelayProfile`, `EmailsProfile`, and `PhoneProfile`? If so, are those separate tables or combined using [abstract base classes](https://docs.djangoproject.com/en/4.2/topics/db/models/#abstract-base-classes)?
   - Which app should have `RegisteredSubdomain`?
   - Which app should have `AbuseMetrics`?

CircleCI does all of the validation that this works. The migrations tests are especially important. If you want to verify they run no SQL:

```sh
./manage.py sqlmigrate privaterelay 0010
./manage.py sqlmigrate emails 0062
```
both return:
```sql
BEGIN;
--
-- Custom state/database change combination
--
-- (no-op)
COMMIT;
```

# Previous PRs

The initial spike was broken up into smaller, focused PRs. Thanks to @groovecoder and @say-yawn for the reviews and making this a better change.

* [x] #4798 - Prefer `user.profile` to other methods of getting the profile
* [x] #4799 - Move exceptions and validators to own files
* [x] #4801 - Use APIClient in IQ tests, and test without DB transactions
* [x] #4802 - Move test user creation to test utils file
* [x] #4822 - Optimize bad word checks, testing if it makes a significant difference
